Add GRPC callback interceptor and correlation IDs to respective API and GRPC handlers

mnafees opened the pull request:
# Description

API and GRPC handlers should be able to "emit" a correlation ID which can be used for various purposes down the execution pipeline.

## Type of change

<!-- Please delete options that are not relevant. -->

- [x] New feature (non-breaking change which adds functionality)

dependabot[bot] committed:
chore(deps): bump hatchet-sdk in /examples/python/quickstart (#2070)

Bumps hatchet-sdk from 1.16.3 to 1.16.4.

---
updated-dependencies:
- dependency-name: hatchet-sdk
  dependency-version: 1.16.4
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): bump google.golang.org/api from 0.243.0 to 0.244.0 (#2071)

Bumps [google.golang.org/api](https://github.com/googleapis/google-api-go-client) from 0.243.0 to 0.244.0.
- [Release notes](https://github.com/googleapis/google-api-go-client/releases)
- [Changelog](https://github.com/googleapis/google-api-go-client/blob/main/CHANGES.md)
- [Commits](https://github.com/googleapis/google-api-go-client/compare/v0.243.0...v0.244.0)

---
updated-dependencies:
- dependency-name: google.golang.org/api
  dependency-version: 0.244.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

mnafees committed:
add grpc callback interceptor

mnafees committed:
add correlation id to more endpoints

mnafees committed:
fix string interpolation payment methods (#2072)

mrkaye97 committed:
hotfix: empty scope in OLAP replication (#2068)

mnafees committed:
fix lint

mnafees committed:
update comment

grutt committed:
feat: activity detection (#2055)

* feat: activity detection

* address comments

abelanger5 commented on api/v1/server/handlers/v1/workflow-runs/trigger.go:
```diff
@@ -122,6 +122,8 @@ func (t *V1WorkflowRunsService) V1WorkflowRunCreate(ctx echo.Context, request ge
 		return nil, err
 	}
 
+	ctx.Set("correlationId", sqlchelpers.UUIDToStr(rawWorkflowRun.WorkflowRun.ExternalID))
```
I'd be strongly in favor of searching for a specific key in the additional metadata as well, and setting that as the correlation ID if it exists. Note that I think correlation ID and resource ID are distinct and we should probably store both

dependabot[bot] committed:
chore(deps): bump github.com/prometheus/client_golang (#2074)

Bumps [github.com/prometheus/client_golang](https://github.com/prometheus/client_golang) from 1.22.0 to 1.23.0.
- [Release notes](https://github.com/prometheus/client_golang/releases)
- [Changelog](https://github.com/prometheus/client_golang/blob/v1.23.0/CHANGELOG.md)
- [Commits](https://github.com/prometheus/client_golang/compare/v1.22.0...v1.23.0)

---
updated-dependencies:
- dependency-name: github.com/prometheus/client_golang
  dependency-version: 1.23.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): bump github.com/getsentry/sentry-go from 0.34.1 to 0.35.0 (#2075)

Bumps [github.com/getsentry/sentry-go](https://github.com/getsentry/sentry-go) from 0.34.1 to 0.35.0.
- [Release notes](https://github.com/getsentry/sentry-go/releases)
- [Changelog](https://github.com/getsentry/sentry-go/blob/master/CHANGELOG.md)
- [Commits](https://github.com/getsentry/sentry-go/compare/v0.34.1...v0.35.0)

---
updated-dependencies:
- dependency-name: github.com/getsentry/sentry-go
  dependency-version: 0.35.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

mnafees committed:
Merge branch 'main' into nafees/handlers-set-correlation-id

mnafees committed:
add resource id and type

mnafees committed:
update grpc callback middleware

mnafees committed:
fix v0 trigger

abelanger5 commented on api/v1/server/handlers/v1/workflow-runs/trigger.go:
```diff
@@ -122,6 +122,16 @@ func (t *V1WorkflowRunsService) V1WorkflowRunCreate(ctx echo.Context, request ge
 	}
 
+	if request.Body.AdditionalMetadata != nil {
+		correlationId, ok := (*request.Body.AdditionalMetadata)["correlationId"].(string)
```
Probably safer to do:

```go
if correlationIdInterface, exists := (*request.Body.AdditionalMetadata)["correlationId"]; exists {
        if correlationId, ok := correlationIdInterface.(string); ok {

        }
    }
```

For all of these - since additional metadata is a `map[string]interface{}` and some old SDKs were technically able to send JSON as the interface value

abelanger5 commented on internal/services/admin/server.go:
```diff
@@ -167,6 +175,12 @@ func (a *AdminServiceImpl) bulkTriggerWorkflowV0(ctx context.Context, req *contr
+		if req.Workflows[i].AdditionalMetadata != nil {
+			corrId = extractCorrelationId(*req.Workflows[i].AdditionalMetadata)
+		}
+		grpcmiddleware.TriggerCallback(ctx, workflowRunId, "workflow-run", corrId)
```
Why aren't we setting these in `ctx` in this case?

abelanger5 commented on api/v1/server/handlers/workflows/create_scheduled.go:
```diff
@@ -41,6 +41,14 @@ func (t *WorkflowService) ScheduledWorkflowRunCreate(ctx echo.Context, request g
 		), nil
 	}
 
+	correlationId, ok := request.Body.AdditionalMetadata["correlationId"].(string)
```
I'm also worried about panics here, there doesn't seem to be enough guarding?

mnafees committed:
use constants

mnafees committed:
fix values

mnafees committed:
use constants

abelanger5 commented on api/v1/server/handlers/api-tokens/revoke.go:
```diff
@@ -24,5 +25,8 @@ func (a *APITokenService) ApiTokenUpdateRevoke(ctx echo.Context, request gen.Api
 	}
 
+	ctx.Set(string(constants.ResourceIdKey), apiToken.ID.String())
+	ctx.Set(string(constants.ResourceTypeKey), string(constants.ResourceTypeApiToken))
```
```suggestion
	ctx.Set(string(constants.ResourceTypeKey), constants.ResourceTypeApiToken.String())
```

thought that was the point of adding the string method?

abelanger5 commented on api/v1/server/handlers/workflows/create_scheduled.go:
```diff
@@ -41,6 +42,17 @@ func (t *WorkflowService) ScheduledWorkflowRunCreate(ctx echo.Context, request g
+	}
+
+	ctx.Set(string(constants.ResourceIdKey), scheduled.ID.String())
+	ctx.Set(string(constants.ResourceTypeKey), "scheduled-workflow")
```
missed constant?

mnafees committed:
use string declared method
